Introduce event-sourcing style in the HeadLogic

pgrange opened the pull request:
To prepare for event-sourced persistency we first revamp the HeadlLogic to make it event-sourced.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Introduce StateChanged events and aggregate function

We use a dummy StateReplaced to have a backward compatible way to inject
a full state update through the new interfaces.

abailly committed:
Introduce HeadInitialized event

pgrange committed:
Introduce CommittedUTxO event

ch1bo committed:
Change 'step' to aggregate the HeadState

This also does not return a state and instead introduces another
getState function to access the state for assertions.

ffakenz committed:
Introduce TransactionAppliedToLocalUTxO event to ReqTx handler

Refactoring the assertNewState into step + getState makes use of
aggregate internally and moves the test code a bit away of using
'collectState'.

pgrange committed:
Introduce SnapshotRequestDecided event

v0d1ch committed:
Introduce SnapshotAppliedToLocalUTxO event

pgrange committed:
Introduce HeadAborted event

pgrange committed:
Introduced HeadOpened event

pgrange committed:
Introduce InvalidTransactionReceived event

pgrange committed:
Refactor: TransactionReceived emitted everytime we receive a transaction

pgrange committed:
Introduce SnapshotHasBeenConfirmedEvent

We don't call the event SnaphotConfirmed because of a name collision
with Server. Should probably qualify Server at some point.

In AckSn:
* we use SnapshotHasBeenConfirmedEvent to notify that a snapshot has been confirmed
* we use SnapshotRequestDecided to notify we requested a snapshot
* we use SnapshotHasBeenConfirmed when the receiving party is not the leader or its
localTxs are null.

ffakenz committed:
Extract aggregateState logic from spec so it is used by the Node as well.

The goal is to keep the ordering constraints of Combined only inside the HeadLogic.

abailly committed:
Add event for partially signed snapshots

abailly committed:
Handle closed event

v0d1ch committed:
Introduce HeadFannedOut event

ffakenz committed:
Introduce HeadIsReadyToFanout event

pgrange committed:
Introduce ChainRolledBack event

pgrange committed:
Introduced TickObserved event

pgrange committed:
All events introduced, removind StateReplaced

pgrange committed:
Reduce data in SnapshoRequested event

The pruning of the transactions can be done in the
events aggregation function instead of in the update
function so we do not need to carry transactions and
utxo in the event anymore

pgrange committed:
Rename SnapshotHasBeenConfirmed to SnapshotConfirmed

This really hilight the symetry between events and server outputs.

pgrange committed:
Only include the just received party signature in PartySignedSnapshot

Make it more clear what the event is about. Move the agregation to the
aggregate and not in the update anymore.

pgrange committed:
Review comment: remove some TODOs

ffakenz commented on hydra-node/src/Hydra/HeadLogic/Error.hs:
```diff
@@ -47,4 +48,4 @@ deriving instance (ToJSON (TxIdType tx)) => ToJSON (RequirementFailure tx)
 
 instance Arbitrary (TxIdType tx) => Arbitrary (RequirementFailure tx) where
-  arbitrary = genericArbitrary
+  arbitrary = genericArbitrary
```
nitpick: add new line at EOF

ffakenz commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -116,7 +116,10 @@ stepHydraNode tracer node = do
+    StateChanged sc -> do
+      -- TODO: We should not need to query the head state here
+      s <- atomically queryHeadState
+      save $ aggregate ledger s sc
```
idk if here or in `processNextEvent`, but I think we should log a warn message in case, after processing an event, the state remains the same.

basically I want to know how many times we fall under the `_otherState` case during state `aggregate`, as imo those events are invalid.

ch1bo replied:
Interesting idea.

If your assumption - given an event sequence is valid, we will never _not update_ the state after a call to `aggregate` - is true, then we could write a `validateEvents` function which does just that and call it on the events whenever we want!

So not only when processing the events one by one, also after the fact or fail startup if that is the case (would be quite restrictive).

ffakenz reviewed (approved):
Besides missing TODOs to tackle, this LGTM.

Just as a suggestion, I would keep the Spec state changes comments on the `aggregate`.

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -116,7 +116,10 @@ stepHydraNode tracer node = do
+      -- TODO: We should not need to query the head state here
+      s <- atomically queryHeadState
+      save $ aggregate ledger s sc
     Effects _ -> pure ()
```
Maybe not related to this PR, but I would have expected the effect processing next to the state change persisting, next to the re-enqueuing. i.e. processEffects to be here and not done in a separate `mapM` (on line 110)

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -190,6 +185,7 @@ data NodeState tx m = NodeState
   }
 
 -- | Initialize a new 'NodeState'.
+-- TODO: We do not need a TVar to keep the HeadState. REMOVE.
```
Do we want to do anything about this now?

I see two alternatives:
 * Explicitly handing the `HeadState tx` around between invocations of `stepHydraNode`, changing it's signature to `stepHydraNode :: HydraNode tx m -> HydraState tx -> m (HydraState tx)`
 * Making the monad we run in stateful. In mtl-style that would be `stepHydraNode :: MonadState (HydraState tx) m => HydraNode tx m -> m ()`, which makes us to use `StateT (HeadState tx) IO` or a `ReaderT (TVar (HeadState tx)) IO` (or similar).
 
Alternative 1 is more explicit, a bit more verbose, and strictly less powerful (maybe good) as it's obviously not concurrently useful.
Alternative 2 would have very similar capabilities and complexity (instead of creating the handle you would need to run the monad transformer stack), but a departure from the handle pattern which we use throughout, and I would prefer consistency.

ch1bo replied:
CC @pgrange for later reference maybe

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -531,43 +469,18 @@ onOpenNetworkAckSn Environment{party} openState otherParty snapshotSignature sn
-                          , seenSnapshot = LastSeenSnapshot sn
-                          }
-                    )
+                else NoOutcome
```
I think we should avoid and remove `NoOutcome`. It should never be just "NoOutcome".

To make it work here, we can just move the `Combined` combinator into the `then` and have an identify for the else.

Or we create a function which takes a list of `[Outcome]` and combines them into the tree form (or we change the type to have `| Combined [Outcome]`) which would have `[]` as the identify for the else branch or we can use a list comprehension like:

```
Combined
  [ Effects [ClientEffect $ ServerOutput.SnapshotConfirmed headId snapshot multisig]
  , StateChanged SnapshotConfirmed{snapshot, signatures = multisig}
  ] ++
  [ Combined [ StateChanged SnapshotRequestDecided{snapshotNumber = nextSn}
             , Effects [NetworkEffect (ReqSn nextSn (txId <$> localTxs))
             ]
  | isLeader parameters party nextSn && not (null localTxs)
  ]
```

..typing it, it may not be that clear and a `[]` or `id` on the else is preferrable.

v0d1ch replied:
If we are happy to have `Effects []` which is equivalent to `NoOutcome` we could use that and remove the `NoOutcome` constructor.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -802,29 +701,249 @@ update env ledger st ev = case (st, ev) of
+  (Open OpenState{}, OnChainEvent Tick{chainSlot}) ->
+    StateChanged (TickObserved{chainSlot})
   (_, OnChainEvent Tick{}) ->
     NoOutcome
```
Should: this be also `TickObserved`?

v0d1ch replied:
Or perhaps `Effects []`?

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -20,16 +20,17 @@ module Hydra.HeadLogic (
+  aggregate,
+  aggregateState,
+)
+where
```
Should: update the module documentation (and function documentations) to explain how `update` and `aggregate` work, as well as how they correspond to the protocol logic from the specification

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -88,6 +90,7 @@ import Hydra.Ledger (
  )
 import Hydra.Network.Message (Message (..))
 import Hydra.Party (Party (vkey))
+import Hydra.Prelude
```
Must: not move the prelude (an implicit import) into the same block as the other imports (which are explicit)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -251,33 +242,16 @@ onInitialChainCollectTx ::
-
   -- TODO: Do we want to check whether this even matches our local state? For
   -- example, we do expect `null remainingParties` but what happens if it's
   -- untrue?
```
I think this TODO lost it's context over the many changes to this file. Now it's just garbage and we can remove it

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -308,51 +282,33 @@ onOpenNetworkReqTx ::
         | ttl > 0 ->
-            NewState (Open st{coordinatedHeadState = chs})
+            StateChanged TransactionReceived{tx}
               `Combined` Wait (WaitOnNotApplicableTx err)
```
This is a bit surprising now here. I got confused by the fact that `TransactionReceived` was used on both branches (here and on line 300). Also, the `TransactionReceived` event will be emitted on each invocation when ttl is still not 0.

Which leads me to the question, would we want to avoid that? Wouldn't a more meaningful event stream be: `TransactionReceived`, `TransactionNotApplicable`, `TransactionNotApplicable`, ..., `TxInvalid`

Not sure what to make out of this yet..

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -413,8 +369,12 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxIds =
     -- Spec: wait s̅ = ŝ
     waitNoSnapshotInFlight $
+      -- Spec: wait ∀h ∈ Treq : (h, ·) ∈ Tall
+
```
Must: remove duplicate comments

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -308,51 +282,33 @@ onOpenNetworkReqTx ::
   Outcome tx
-onOpenNetworkReqTx env ledger st ttl tx = do
-  -- Spec: Tall ← ̂Tall ∪ { (hash(tx), tx) }
-  let chs' = coordinatedHeadState{allTxs = allTxs <> fromList [(txId tx, tx)]}
```
Should: keep the spec reference and the fact that we want to update the state via the `TransactionReceived` state change event here.

I think writing something like on line 323 (old code) could work?

```
-- Spec: Tall ← ̂Tall ∪ { (hash(tx), tx) }
(StateChanged TransactionReceived{tx} `Combined`) $
  -- Spec: wait L̂ ◦ tx ≠ ⊥ combined with L̂ ← L̂ ◦ tx
  waitApplyTx $ \utxo' ->
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -425,23 +385,13 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxIds =
+              -- TODO: unclear where to put these comments
+              -- Spec: for loop which updates T̂ and L̂
+              -- TODO: where to put this spec comment now?
+              -- Spec: Tall ← {tx | ∀tx ∈ Tall : tx ∉ Treq }
```
A general note or here that the aggregation of `SnapshotRequest` will perform these two sections in the spec would work IMO

v0d1ch replied:
I would keep the `pruneTransactions` where it was here and not move it to the new `aggregate` function. `aggregate` then would not depend on a ledger and would contain only state changes with `Map.insert/delete` here and there and in the end we would not need to update the spec at all.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -655,7 +565,7 @@ onOpenChainCloseTx ::
   Outcome tx
 onOpenChainCloseTx openState newChainState closedSnapshotNumber contestationDeadline =
-  NewState closedState
+  StateChanged (HeadClosed newChainState contestationDeadline)
```
Should: prefer record syntax

This allows re-ordering fields with ease

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -595,41 +625,21 @@ data StepState tx = StepState
+  (MonadState (StepState tx) m, IsChainState tx) =>
   Event tx ->
-  m (HeadState tx)
+  m (Outcome tx)
```
There was a refactoring introducing `getState` to precisely not have this return value and be able to not turn off the unused do binds warning.

Should: remove the need for this return value, have two variants or `void` invocations without the warning disabled

v0d1ch replied:
We thought it is not that important since this is test code. We could also add the `Outcome` as part of the `StepState` and avoid these warnings.

ch1bo reviewed (changes requested):
Some clean up on comments, imports, and maybe the tick observed event should be emitted in all states.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -802,29 +701,249 @@ update env ledger st ev = case (st, ev) of
+        pruneTransactions utxo = do
+          foldr go ([], utxo) localTxs
+         where
+          go tx (txs, u) =
```
We had it the other way and moved it here to keep the events minimal in what they contain. i.e. persist only the minimum of what is required to reconstruct the state.

> Events represent "past" things that transform the state directly

But having a `StateReplaced` event is undesirable still, no? So I thought we should be keeping events small and to the point.

> The ledger is a complex thing to build, it requires parameters which might not be trivial to retrieve and might even change over time within a head

Even more a reason to keep it in the aggregate IMO as those new / current parameters would be in the state if they can change

> The main advantage of event sourcing is that it provides a stream of "raw" data for downstream consumers, possibly even with vastly different usage,

Correct, and not all downstream users (aggregates, policies) have a notion of `localUTxO` or `localTxs`. So to them it would even be noise to keep these things in the event?

> events should be self-contained enough to provide all the data needed to reconstruct the state

Why would you think this is not the case anymore?

v0d1ch replied:
I also prefer keeping the ledger out of `aggregate` function. Then `aggregate` really doesn't depend on anything special and only does some data shuffling between different records.

ch1bo replied:
We did it this way now again.

ffakenz committed:
Improve haddock for update function.

v0d1ch reviewed (changes requested):
Nice chunk of work guys! The main concern I have is keeping the ledger out of `aggregate` function which would simplify things.

ch1bo committed:
Move pruneTransactions back into update and extend SnapshotRequested event

We want to avoid requiring a ledger to interpret the StateChanged events
to allow for simpler consumption downstream (maybe even in non-haskell
contexts).

ch1bo committed:
Export all identifiers in Hydra.HeadLogic and remove a TODO

ch1bo committed:
Update module documentation of HeadLogic

ch1bo committed:
Move the TransactionReceived outcome out of waitApplyTx

This way it is more clear that this event will always be the outcome of
a ReqTx being processed.

ch1bo committed:
Always produce Tick state change event irrespective of state

This way, any downstream code of the events can decide whether they want
to consume that event or not.

ch1bo committed:
Remove NoOutcome and make Outcome a Semigroup

The NoOutcome identity is not needed and now all update outcomes must be
any of the other data constructors.

Also make Combine the mappend of Outcome to make the code a bit easier
to read.

ch1bo committed:
Use record syntax in constructing HeadClosed

ch1bo reviewed (approved):
With the changes we did in ensemble, I think this is a good step forward now.
